ethportal-api/src/types/content_value/state.rs

```diff
@@ -0,0 +1,285 @@
+        }
+    }
+
+    fn decode(buf: &[u8]) -> Result<Self, ContentValueError> {
+        // Catch any attempt to construct a content value from "0x" improperly.
+        if buf == CONTENT_ABSENT.to_string().as_bytes() {
+            Err(ContentValueError::DecodeAbsentContent)
+        } else if let Ok(value) = TrieNode::from_ssz_bytes(buf) {
+            Ok(Self::TrieNode(value))
+        } else if let Ok(value) = AccountTrieNodeWithProof::from_ssz_bytes(buf) {
+            Ok(Self::AccountTrieNodeWithProof(value))
+        } else if let Ok(value) = ContractStorageTrieNodeWithProof::from_ssz_bytes(buf) {
+            Ok(Self::ContractStorageTrieNodeWithProof(value))
+        } else if let Ok(value) = ContractBytecode::from_ssz_bytes(buf) {
+            Ok(Self::ContractBytecode(value))
+        } else if let Ok(value) = ContractBytecodeWithProof::from_ssz_bytes(buf) {
+            Ok(Self::ContractBytecodeWithProof(value))
+        } else {
+            Err(ContentValueError::UnknownContent {
+                bytes: hex_encode(buf),
+                network: "state".to_string(),
+            })
+        }
+    }
+}
+
+/// A content value type, used when retriving a trie node.
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct TrieNode {
+    pub node: EncodedTrieNode,
+}
+
+/// A content value type, used when offering a trie node from the account trie.
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct AccountTrieNodeWithProof {
+    /// An account trie node with a proof.
+    pub proof: AccountTrieWitness,
+    /// A block at which the proof is anchored.
+    pub block_hash: H256,
+}
+
+/// A content value type, used when offering a trie node from the contract storage trie.
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct ContractStorageTrieNodeWithProof {
+    /// A contract storage trie node with a proof.
+    pub proof: ContractStorageTrieWitness,
+    /// A block at which the proof is anchored.
+    pub block_hash: H256,
+}
+
+pub type ByteCode = VariableList<u8, typenum::U65536>;
+
+/// A content value type, used when retriving contract's bytecode.
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct ContractBytecode {
+    pub code: ByteCode,
+}
+
+/// A content value type, used when offering contract's bytecode.
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct ContractBytecodeWithProof {
+    /// A contract's bytecode.
+    pub code: ByteCode,
+    /// A proof for the account of the corresponding contract.
+    pub account_proof: AccountTrieWitness,
+    /// A block at which the proof is anchored.
+    pub block_hash: H256,
+}
+
+#[cfg(test)]
+mod test {
+    use std::{fs, str::FromStr};
+
+    use anyhow::Result;
+    use rstest::rstest;
+    use serde_json::Value;
+
+    use crate::{
+        types::state_trie::{nibbles::Nibbles, witness::TrieWitness},
+        utils::bytes::hex_decode,
+    };
+
+    use super::*;
+
```

Comment:
This seems like a good place to explain where the test values were sourced from - usually a github link (that contains some commit hash) is ideal

Reply:
I just updated tests with new test vectors that are being discussed in new [PR](https://github.com/ethereum/portal-network-specs/pull/271). I will wait with merging this PR until the other one is merged and then I will add a link here.